server/src/main/java/org/opensearch/search/startree/StarTreeFilter.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.startree;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.DocIdSetBuilder;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.compositeindex.datacube.startree.index.StarTreeValues;
+import org.opensearch.index.compositeindex.datacube.startree.node.StarTreeNode;
+import org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeUtils;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * Filter operator for star tree data structure.
+ *
+ *  @opensearch.experimental
+ */
+public class StarTreeFilter {
+    private static final Logger logger = LogManager.getLogger(StarTreeFilter.class);
+
+    // private final StarTreeNode starTreeRoot;
+    Map<String, Long> queryMap;
+    DocIdSetBuilder docsWithField;
+    DocIdSetBuilder.BulkAdder adder;
+    // Map<String, DocIdSetIterator> dimValueMap;
+    StarTreeValues starTreeValues;
+    List<Dimension> dimensions;
+
+    public StarTreeFilter(StarTreeValues starTreeAggrStructure, Map<String, Long> predicateEvaluators) {
+        // This filter operator does not support AND/OR/NOT operations as of now.
+        starTreeValues = starTreeAggrStructure;
+        // starTreeRoot = starTreeAggrStructure.getRoot();
+        // dimValueMap = starTreeAggrStructure.getDimensionDocValuesIteratorMap();
+        queryMap = predicateEvaluators != null ? predicateEvaluators : Collections.emptyMap();
+
+        // TODO : this should be the maximum number of doc values
+        docsWithField = new DocIdSetBuilder(Integer.MAX_VALUE);
```

Comment:
Pass in `LeafReaderContext` from `StarTreeQuery` and use `ctx.reader().maxDoc()`.

Reply:
The only concern with that I see that the number when star-tree documents exceed the `ctx.reader().maxDoc()` which we noticed in some runs. 
@bharath-techie Any thoughts?

Reply:
Yes. Maybe we can use star tree's maxdoc (from star tree values) instead of ctx.reader().maxdoc if needed.

But still there is a question of which star tree will you pick in case of multiple star trees ( not an issue for 2.17 but raising this to be future proof )

Reply:
> The only concern with that I see that the number when star-tree documents exceed the ctx.reader().maxDoc() which we noticed in some runs.

That should never happen. The number of documents in a segment (including deleted docs) is (by definition) `maxDoc`. If we have doc IDs that exceed `maxDoc()`, then the index is corrupted.

By allocating a `DocIdSetBuilder` with `Integer.MAX_VALUE`, we end up using an `int[]` up to 16 million docs, then flip to allocating a `FixedBitSet` of size 2.147 billion after that. It's important that the `DocIdSetBuilder` is allocated according to the doc ID space that it uses.